### PR TITLE
Test lib

### PR DESCRIPTION
**Pull request by jamesmaa**
Wrote some tests for `2x/src/lib`

Here is the current status:
image

I'm having trouble writing some of the tests. @nramadas @phil303 do you guys have experience writing tests with the following constraints? 
- Dealing with browser capabilities (e.g. localStorage, cookies)
- Mocking objects/requests

I skipped some tests because they seemed pretty short/obvious

### CODE REVIEW

File: test/lib/classNames.test.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    it('is a function', () => {
+        expect(classNames).to.be.a('function');
+    });
+
+    it('creates correct classname for string types', () => {
+        expect(classNames('hello', 'world', 'baby')).to.equal('hello world baby');
+        expect(classNames('hello')).to.equal('hello');
+    });
+
+    it('allows duplicate css classes', () => {
```
Comment (phil303) [inline test/lib/classNames.test.js]: :no_bell: arguably that's an oversight on my part and not a feature :)

File: test/lib/classNames.test.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        expect(classNames(['hello', 'world', 'baby'])).to.equal('hello world baby');
+    });
+
+    it('creates correct classname for object types', () => {
+        expect(classNames({ baz: false, bob: true })).to.equal('bob');
+    });
+
+    it('does not support nested dictionaries. Only allows truthy values', () => {
+        const d = { baz: true };
+        d.bob = d;
```
Comment (phil303) [inline test/lib/classNames.test.js]: this test makes me a weeeee bit nervous. If the day comes we decide for it to recurse then this is just going to overflow the stack. Maybe just make this another object?

File: test/lib/cleanObject.test.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+const expect = chai.expect;
+
+chai.use(sinonChai);
+
+describe('lib: cleanObject', () => {
+    it('is a function', () => {
+        expect(cleanObject).to.be.a('function');
+    });
+
+    const obj = {foo: null, bar: undefined, baz: true};
```
Comment (phil303) [inline test/lib/cleanObject.test.js]: style nit, a space around the curlies

File: test/lib/cleanObject.test.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    it('removes null and undefined values', () => {
+        expect(cleanObject(obj)).to.eql({baz: true});
```
Comment (phil303) [inline test/lib/cleanObject.test.js]: same here

File: test/lib/RingStack.test.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        expect(stack.head).to.equal(3);
+        stack.push(7);
+        expect(stack.values()).to.eql([7, 6, 5, 4]);
+        expect(stack.head).to.equal(0);
+        stack.push(8);
+        expect(stack.values()).to.eql([8, 7, 6, 5]);
+        expect(stack.head).to.equal(1);
+
+    })
+});
```
Comment (phil303) [inline test/lib/RingStack.test.js]: style nit, if you're using sublime text, there's a preference called something like "ensure newline at EOF", mind turning that on?

**Comment by phil303** [comment]
This is awesome work! Thanks for taking it on. There are certain ways to mimic _most_ of the dom in node, namely the library [jsdom](https://github.com/tmpvar/jsdom) and its compatriots. But let's chat/evaluate whether we need to use it here. We might be able to get away with not pulling in another library.

:nail_care:

**Comment by jamesmaa** [comment]
Cool thanks for the comments, I found the Sublime preference for `ensure_newline_at_eof_on_save`

With some of the tests, I wrote tests that froze the behavior of the functions. e.g. for classNames, I knew that the function would allow duplicate class names even though it was not an intended function. I figured that it would be better to use these tests in an expository manner to document how functions behave, so devs in the future could change the function/tests if they wish to alter the behavior. 

I'm not sure if that's the purpose of these tests, since the libs don't have documentation themselves.

**Comment by phil303** [comment]
I very much got that vibe and I'm totally for it (which is why I no_bell'd that one test). When you say "the libs don't have documentation themselves", I'm curious what you'd like to see in that domain. If you've heard of [jsdocs](https://en.wikipedia.org/wiki/JSDoc), do you think that'd be helpful here? Or are you looking for more thorough, more prose-like documentation on a wiki of some sort?

**Comment by jamesmaa** [comment]
I'm leaning more towards "simple -> better" so I think as long 

I was working with the assumption that if you don't have documentation, your code becomes your documentation. So as long I don't have to thoroughly read the code to understand intended behavior, I'm fine with either. jsdocs seems like the easier solution, and I don't think we even need to be that thorough since most of the functions I'm testing are pretty simple anyways.

**Comment by phil303** [comment]
:fish:! make sure to rebase

**Pull request closed without merging**